qiskit/algorithms/optimizers/suffix_averaging.py

```diff
@@ -0,0 +1,144 @@
+        Stochastic Gradient Line Bayesian Optimization
+        for Efficient Noise-Robust Optimization of Parameterized Quantum Circuits.
+        arXiv preprint arXiv:2111.07952.
+    """
+
+    def __init__(self, optimizer: Optimizer, n_params_suffix: int = 50) -> None:
+        """
+        Args:
+            optimizer: The optimizer used for optimizing parameterized quantum circuits.
+            n_params_suffix: The number of circuit parameters for taking the suffix average.
+        """
+
+        self._n_params_suffix = n_params_suffix
+        self._optimizer = optimizer
+
+        self._circ_params: list[POINT] = []
+
+        super().__init__()
+
+    def get_support_level(self):
+        """Return support level dictionary"""
+        return self._optimizer.get_support_level()
+
+    def _return_suffix_average(self) -> np.ndarray:
+
+        if isinstance(self._optimizer, NFT):
+            n_params = 1 if isinstance(self._circ_params[0], float) else len(self._circ_params[0])
+            n_repitition = int(len(self._circ_params) / n_params)
+            if n_repitition < self._n_params_suffix:
+                warnings.warn("The total number of iterations is less than n_params_suffix.")
+                averaged_param = np.mean(
+                    self._circ_params[: -n_params * n_repitition : -n_params], axis=0
+                )
+                return averaged_param
+
+            averaged_param = np.mean(
+                self._circ_params[: -n_params * self._n_params_suffix : -n_params], axis=0
+            )
+
+        else:
+            n_iterates = len(self._circ_params)
+
+            if n_iterates < self._n_params_suffix:
+                warnings.warn("The total number of iterations is less than n_params_suffix.")
+                averaged_param = np.mean(self._circ_params, axis=0)
+                return averaged_param
+
+            averaged_param = np.mean(self._circ_params[-self._n_params_suffix :], axis=0)
+
+        return averaged_param
+
+    @property
+    def settings(self) -> Dict[str, Any]:
+        return {"optimizer": self._optimizer, "n_params_suffix": self._n_params_suffix}
+
+    def minimize(
+        self,
+        fun: Callable[[POINT], float],
+        x0: POINT,
+        jac: Optional[Callable[[POINT], POINT]] = None,
+        bounds: Optional[List[Tuple[float, float]]] = None,
+    ) -> OptimizerResult:
+
+        original_callback = self._optimizer.callback
+        if isinstance(self._optimizer, (SPSA, QNSPSA)):
+
+            def callback(
+                nfev: int,
+                x_next: POINT,
+                fx_next: float,
+                stepsize: np.floating,
+                is_accepted: bool,
+            ):
+                self._circ_params.append(x_next)
+                if original_callback is not None:
+                    original_callback(nfev, x_next, fx_next, stepsize, is_accepted)
+
+        elif isinstance(self._optimizer, GradientDescent):
```

Comment:
With a unified signature for the callbacks we don't need any special cases, right? 🙂